Content.Server/Chat/V2/Systems/ChatAttemptHandlerSystem.cs

```diff
@@ -0,0 +1,213 @@
+    }
+
+    private bool IsDeadChatValid(AttemptOutOfCharacterChatEvent ev, EntitySessionEventArgs args, EntityUid entityUid)
+    {
+        var isAdmin = _admin.IsAdmin(entityUid);
+
+        // Non-admins can only talk on dead chat if they're a ghost or currently dead.
+        if (!isAdmin && !HasComp<GhostComponent>(entityUid) && !_mobState.IsDead(entityUid))
+        {
+            OnFailedMessage(new OutOfCharacterChatFailed(ev.Context, ev.Sender, _deadChatFailed), args);
+
+            return false;
+        }
+        // ReSharper disable once InvertIf
+        if (!ValidateMessage(entityUid, ev, args.SenderSession, out var reason))
+        {
+            OnFailedMessage(new OutOfCharacterChatFailed(ev.Context, ev.Sender, reason), args);
+
+            return false;
+        }
+
+        return true;
+    }
+
+    public void OnFailedMessage(ChatFailedEvent ev, EntitySessionEventArgs args)
+    {
+        RaiseNetworkEvent(ev, args.SenderSession);
+    }
+
+    /// <summary>
+    /// Validates messages. Return true means the message is valid.
+    /// </summary>
+    private bool ValidateMessage<T>(EntityUid entityUid, T ev, ICommonSession player, out string reason) where T : ChatAttemptEvent
+    {
+        // This check is simple, so we don't need to raise an event for it.
+        if (player.AttachedEntity != null || player.AttachedEntity != entityUid)
+        {
+            reason = _entityNotOwnedBySender;
+
+            return false;
+        }
+
+        // Raise both the general and specific ChatValidationEvents. This allows for general
+        var validate = new ChatValidationEvent<ChatAttemptEvent>(ev);
+        RaiseLocalEvent(entityUid, validate);
+
+        if (validate.IsCancelled)
+        {
+            reason = validate.Reason;
+
+            return false;
+        }
+
+        var validateT = new ChatValidationEvent<T>(ev);
+        RaiseLocalEvent(entityUid, validateT);
+
+        if (validate.IsCancelled)
+        {
+            reason = validate.Reason;
+
+            return false;
+        }
+
+        reason = "";
+
+        return true;
+    }
+
+    /// <summary>
+    /// Validates messages which depend on a sentinel component to be legal.
+    /// </summary>
+    private bool ValidateMessage<T1, T2>(
+        T1 ev,
+        ICommonSession player,
+        out string reason,
+        [NotNullWhen(true)] out T2? comp
+    ) where T1 : ChatAttemptEvent where T2 : IComponent
+    {
+        comp = default;
+
+        var entityUid = GetEntity(ev.Sender);
+
+        if (!ValidateMessage(entityUid, ev, player, out reason))
+        {
+            return false;
+        }
+
+        if (TryComp(entityUid, out comp))
+            return true;
+
+        reason = _missingRequiredComponent;
+
+        return false;
+    }
+
+    /// <summary>
+    /// Sanitizes messages. The return string is sanitized.
+    /// </summary>
+    private void SanitizeMessage<T>(T evt) where T : IChatEvent
```

Comment:
Note there are no sanitization handlers at the moment. I leave this to @ShadowCommander, @Chief-Engineer and co to sort out as this is an admin topic.